### ISSUE TITLE
Inconsistencies with available assignment operators

### ISSUE THREAD

**Issue by lukebemish** [JuliaLang/julia#56821]
Most of the basic "arithmetic" operators support assignment operators, where something like `a += 5` is equivalent to `a = a + 5`. However, the presence of these seems to be somewhat inconsistent and a couple useful use-cases are missing. One use case I've ran into recently, when working with immutable structures, is wanting to update the value of a variable in-place in a similar fashion, but when applying arbitrary functions. For instance, something like
```julia
a = f(a)
```
or even
```julia
a = f(g(a))
```
The `|>` operator makes the latter a bit nicer to express, as
```julia
a = a |> g |> f
```
But it would be convenient to be able to do something like
```julia
a |>= g ∘ f
```
or the like instead. Furthermore, there are inconsistencies even among the more "normal" infix operators in terms of which support assignment operators; while `⊼`, `⊽`, and `⊻` are all infix operators (giving bitwise nand, nor, and xor respectively), only `⊻=` is a valid assignment operator.

**Comment by Keno** [comment]
> Furthermore, there are inconsistencies even among the more "normal" infix operators in terms of which support assignment operators; while `⊼`, `⊽`, and `⊻` are all infix operators (giving bitwise nand, nor, and xor respectively), only `⊻=` is a valid assignment operator, which feels wacky.

These are probably just missing from the list in the parser, might as well have them.

**Comment by lukebemish** [comment]
Is there a reason not to just have assignment operators for every infix operator where it wouldn't cause ambiguity? (That is, not `<`). The current list feels a bit arbitrary to be honest.

**Comment by lukebemish** [comment]
Simplest in my mind would be to make an equivalent assignment operator for every operator that's an actual function, so excluding"control flow" that's effectively not evaluated like an infix operator (`&&`), comparison operators in general (since (a) if it returns a boolean, in-place updates aren't useful or intuitive, and (b) they're likely to be confusing), or those operators for which the assignment operator might have some other implied meaning (`:`, for instance). At any rate, you're right that it's more art than science -- hence why I think the best approach is just to provide an assignment operator for every operator where doing so isn't immediately useless or problematic. Otherwise, providing them for only a small subset is rather arbitrary -- especially when that subset already includes operators many folks may not recognize right off the bat, or use at all, like `⊻`.

**Comment by JeffBezanson** [comment]
> while `⊼`, `⊽`, and `⊻` are all infix operators (giving bitwise nand, nor, and xor respectively), only `⊻=` is a valid assignment operator

That looks like an oversight; we should just add them.

**Comment by StefanKarpinski** [comment]
Would also love to have `∨=` and `∧=` as well as `∪=` and `∩=`.

**Comment by adienes** [comment]
I agree all of those are super reasonable; my being a hater extends only to `|>` and all the comparison operators

**Comment by LilithHafner** [comment]
Good first issue if you want to work on the parser but not if you want to write ordinary Julia code.

**Comment by LilithHafner** [comment]
Triage was broadly in support of making infix operators such as  `⊼`, `⊽`, `⊻`, `∨`, `∧` , `∪`, and `∩` also have updating variants, but didn't talk about `|>`, specifically.

Personally, I'm not a fan of `|>=`.

For someone making a PR, I'd recommend finding the list of infix operators in the parser and supporting all of them that don't seem confusing.

I'll put the triage label back on to talk about `|>=`, specifically.

**Comment by tecosaur** [comment]
Related: #46371, #47899

**Comment by mbauman** [comment]
A more principled approach here could be to automatically support assignment-ified versions of all operators in certain precedence groups. 

https://github.com/JuliaLang/julia/blob/316f0fe0b0e454093b44095bb99b273d4158bdc0/src/julia-parser.scm#L22-L30

Grabbing this swatch of operators (from plus through power) would both cover all current assignment-ifications and also look to all be quite unambiguous.   Those are also the most arithmeticky sections.

It would also give a fairly concise rationale for why `a := b` or `a <= b` don't mean `a = a:b` or `a = a < b` — those aren't plus/times/shift or power-like.

**Comment by adienes** [comment]
dup of https://github.com/JuliaLang/julia/issues/40079 but this one seems to be now the canonical issue thread

**Comment by StefanKarpinski** [comment]
Triage thinks we should do #40079 and possibly add the two pipe precedences, which would address this issue too.
